xds/internal/server/rds_handler_test.go

```diff
@@ -20,325 +20,391 @@ package server
-	// this should write to the update channel to send to the listener.
-	fakeClient.InvokeWatchRouteConfigCallback(route1, rdsUpdate, nil)
-	rhuWant := map[string]xdsresource.RouteConfigUpdate{route1: rdsUpdate}
-	select {
-	case rhu := <-ch:
-		if diff := cmp.Diff(rhu.updates, rhuWant); diff != "" {
-			t.Fatalf("got unexpected route update, diff (-got, +want): %v", diff)
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for update from update channel.")
+	// Verify that the update is pushed to the handler's update channel.
+	wantUpdate := rdsHandlerUpdate{
+		updates: map[string]xdsresource.RouteConfigUpdate{
+			route1: defaultRouteConfigUpdate,
+			route2: defaultRouteConfigUpdate,
+		},
 	}
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
 
+	// Update the handler to watch only one of the two previous routes.
+	rh.updateRouteNamesToWatch(map[string]bool{route1: true})
+
+	// Verify that the other route is no longer requested.
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{route1})
+
+	// Verify that an update is pushed with only one route.
+	wantUpdate = rdsHandlerUpdate{updates: map[string]xdsresource.RouteConfigUpdate{route1: defaultRouteConfigUpdate}}
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
+
+	// Close the rds handler and verify that the watch is canceled.
 	rh.close()
-	routeNameDeleted, err = fakeClient.WaitForCancelRouteConfigWatch(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.CancelRDS failed with error: %v", err)
-	}
-	if routeNameDeleted != route1 {
-		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v", routeNameDeleted, route1)
-	}
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{})
 }
 
-// TestSuccessCaseTwoUpdatesAddAndDeleteRoute tests the case where the rds
+// Tests the case where the rds
 // handler receives an update with two routes, and then receives an update with
 // two routes, one previously there and one added (i.e. 12 -> 23). This should
 // cause the route that is no longer there to be deleted and cancelled, and the
 // route that was added should have a watch started for it.
-func (s) TestSuccessCaseTwoUpdatesAddAndDeleteRoute(t *testing.T) {
-	rh, fakeClient, ch := setupTests()
+func (s) TestRDSHandler_SuccessCaseTwoUpdatesAddAndDeleteRoute(t *testing.T) {
+	mgmtServer, nodeID, resourceNamesCh, xdsC := setupForRDSHandlerTests(t)
 
+	// Create an rds handler and give it two routes to watch.
+	updateCh := make(chan rdsHandlerUpdate, 1)
+	rh := newRDSHandler(xdsC, updateCh)
 	rh.updateRouteNamesToWatch(map[string]bool{route1: true, route2: true})
 
+	// Verify that the given routes are requested.
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	if err := waitForFuncWithNames(ctx, fakeClient.WaitForWatchRouteConfig, route1, route2); err != nil {
-		t.Fatalf("Error while waiting for names: %v", err)
-	}
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{route1, route2})
```

Comment:
Optional: this seems pretty common to many tests, can maybe pull this out to helper.

Reply:
I don't see what is common that can be pulled into a helper. Not all tests start with two routes. Some start with one. And I think this is a core portion of the test: creating the handler, giving it one or more names to watch and verifying that it sends RDS requests for the same. Pulling that out into a helper isn't very helpful in my opinion since the reader of the test will not be able to glance over that detail.